cpp/src/arrow/util/future.cc

```diff
@@ -243,6 +253,12 @@ class ConcreteFutureImpl : public FutureImpl {
       }
     }
     cv_.notify_all();
+
+    // run callbacks
+    for (auto&& callback : callbacks_) {
```

Comment:
I'm not sure why this isn't safe. During iteration the vector won't be dropped (since we have at least the reference held by the thread calling `MarkFinished`) or mutated (since attempting to add a callback to a finished future will immediately evaluate that callback)

Reply:
Because you're accessing an object that's mutated (potentially) from another thread in `AddCallback`.
Ideally there should be a multi-threaded stress test for adding callbacks to a future.

Reply:
I agree with Ben.  The vector cannot be mutated because we know by the time we reach the iteration that the future has (safely) been marked finished.  Any `AddCallback` from another thread will not modify the callbacks vector.  I added FutureTest::StressCallback to confirm.